File: go-sdk/runner/runner.go
package runner

import (
	"fmt"
	"time"

	"github.com/konstellation-io/kai-sdk/go-sdk/internal/common"

	"github.com/go-logr/logr"
	"github.com/go-logr/zapr"
	"github.com/konstellation-io/kai-sdk/go-sdk/runner/exit"
	"github.com/konstellation-io/kai-sdk/go-sdk/runner/task"
	"github.com/konstellation-io/kai-sdk/go-sdk/runner/trigger"
	"github.com/nats-io/nats.go"
	"github.com/spf13/viper"
	"go.uber.org/zap"
	"golang.org/x/exp/slices"
)

type Runner struct {
	logger    logr.Logger
	nats      *nats.Conn
	jetstream nats.JetStreamContext
}

func NewRunner() *Runner {
	initializeConfiguration()

	logger := getLogger()

	nc, err := getNatsConnection(logger)
	if err != nil {
		panic(fmt.Errorf("fatal error connecting to NATS: %w", err))
	}

	js, err := getJetStreamConnection(logger, nc)
	if err != nil {
		panic(fmt.Errorf("fatal error connecting to JetStream: %w", err))
	}

	return &Runner{
		logger:    logger,
		nats:      nc,
		jetstream: js,
	}
}

func validateConfig(keys []string) {
	var mandatoryConfigKeys = []string{
		common.ConfigMetadataProductIDKey,
		common.ConfigMetadataWorkflowIDKey,
		common.ConfigMetadataProcessIDKey,
		common.ConfigMetadataVersionIDKey,
		common.ConfigNatsURLKey,
		common.ConfigNatsStreamKey,
		common.ConfigNatsOutputKey,
		common.ConfigCcGlobalBucketKey,
		common.ConfigCcProductBucketKey,
		common.ConfigCcWorkflowBucketKey,
		common.ConfigCcProcessBucketKey,
		common.ConfigMinioEndpointKey,
		common.ConfigMinioClientUserKey,
		common.ConfigMinioClientPasswordKey,
		common.ConfigMinioUseSslKey,
		common.ConfigMinioBucketKey,
		common.ConfigAuthEndpointKey,
		common.ConfigAuthClientKey,
		common.ConfigAuthClientSecretKey,
		common.ConfigAuthRealmKey,
		common.ConfigRedisUsernameKey,
		common.ConfigRedisPasswordKey,
		common.ConfigRedisEndpointKey,
		common.ConfigMeasurementsEndpointKey,
		common.ConfigMeasurementsInsecureKey,
		common.ConfigMeasurementsTimeoutKey,
		common.ConfigMeasurementsMetricsIntervalKey,
	}

	for _, key := range mandatoryConfigKeys {
		if !slices.Contains(keys, key) {
			panic(fmt.Sprintf("missing mandatory configuration key: %s", key))
		}
	}
}

func initializeConfiguration() {
	// Load environment variables
	viper.SetEnvPrefix("KAI")
	viper.AutomaticEnv()

	if viper.IsSet(common.ConfigAppConfigPathKey) {
		viper.AddConfigPath(viper.GetString(common.ConfigAppConfigPathKey))
	}

	viper.SetConfigName("config")
	viper.SetConfigType("yaml")
	viper.AddConfigPath(".")

	err := viper.ReadInConfig() //nolint:ineffassign,staticcheck // err is used in the next if

	viper.SetConfigName("app")
	viper.SetConfigType("yaml")
	viper.AddConfigPath(".")

	if viper.IsSet(common.ConfigAppConfigPathKey) {
		viper.AddConfigPath(viper.GetString(common.ConfigAppConfigPathKey))
	}

	err = viper.MergeInConfig()

	keys := viper.AllKeys()
	if len(keys) == 0 {
		panic(fmt.Errorf("configuration could not be loaded: %w", err))
	}

	validateConfig(keys)

	// Set viper default values
	viper.SetDefault(common.ConfigRunnerSubscriberAckWaitTimeKey, 22*time.Hour)
	viper.SetDefault(common.ConfigRunnerLoggerLevelKey, "InfoLevel")
	viper.SetDefault(common.ConfigRunnerLoggerEncodingKey, "json")
	viper.SetDefault(common.ConfigRunnerLoggerOutputPathsKey, []string{"stdout"})
	viper.SetDefault(common.ConfigRunnerLoggerErrorOutputPathsKey, []string{"stderr"})
}

func getNatsConnection(logger logr.Logger) (*nats.Conn, error) {
	nc, err := nats.Connect(viper.GetString(common.ConfigNatsURLKey))
	if err != nil {
		logger.Error(err, "Error connecting to NATS")
		return nil, err
	}

	return nc, nil
}

func getJetStreamConnection(logger logr.Logger, nc *nats.Conn) (nats.JetStreamContext, error) {
	js, err := nc.JetStream()
	if err != nil {
		logger.Error(err, "Error connecting to JetStream")
		return nil, err
	}

	return js, nil
}

func getLogger() logr.Logger {
	var log logr.Logger

	config := zap.NewProductionConfig()

	logLevel, err := zap.ParseAtomicLevel(viper.GetString(common.ConfigRunnerLoggerLevelKey))
	if err != nil {
		logLevel = zap.NewAtomicLevelAt(zap.InfoLevel)
	}

	config.Level = zap.NewAtomicLevelAt(logLevel.Level())
	config.OutputPaths = viper.GetStringSlice(common.ConfigRunnerLoggerOutputPathsKey)
	config.ErrorOutputPaths = viper.GetStringSlice(common.ConfigRunnerLoggerErrorOutputPathsKey)
	config.Encoding = viper.GetString(common.ConfigRunnerLoggerEncodingKey)

	logger, err := config.Build()
	if err != nil {
		panic("The logger could not be initialized")
	}

	defer logger.Sync() //nolint:errcheck // Ignore error

	log = zapr.NewLogger(logger)

	log.WithName("[RUNNER CONFIG]").V(1).Info(fmt.Sprintf("Logger initialized with level %s", logLevel.String()))

	return log
}

func (rn Runner) TriggerRunner() *trigger.Runner {
	return trigger.NewTriggerRunner(rn.logger, rn.nats, rn.jetstream)
}

func (rn Runner) TaskRunner() *task.Runner {
	return task.NewTaskRunner(rn.logger, rn.nats, rn.jetstream)
}

func (rn Runner) ExitRunner() *exit.Runner {
	return exit.NewExitRunner(rn.logger, rn.nats, rn.jetstream)
}

File: py-sdk/src/runner/runner.py
import sys
from dataclasses import dataclass

from exceptions import FailedLoadingConfigError, JetStreamConnectionError, NATSConnectionError
from exit.exit_runner import ExitRunner
from loguru import logger
from loguru._logger import Logger
from nats.aio.client import Client as NatsClient
from nats.js.client import JetStreamContext
from task.task_runner import TaskRunner
from trigger.trigger_runner import TriggerRunner
from vyper import v


@dataclass
class Runner:
    nc: NatsClient = NatsClient()
    js: JetStreamContext = None
    logger: Logger = None

    def __post_init__(self):
        self.initialize_config()
        self.initialize_logger()

    async def initialize(self):
        try:
            await self.nc.connect(v.get("nats.url"))
        except Exception as e:
            self.logger.error(f"error connecting to nats: {e}")
            raise NATSConnectionError(e)

        try:
            self.js = await self.nc.jetstream()
        except Exception as e:
            self.logger.error(f"error connecting to jetstream: {e}")
            raise JetStreamConnectionError(e)

    def initialize_config(self):
        v.automatic_env()

        if v.is_set("APP_CONFIG_PATH"):
            v.add_config_path(v.get("APP_CONFIG_PATH"))

        v.set_config_name("config")
        v.set_config_type("yaml")
        v.add_config_path(".")

        error = None
        try:
            v.read_in_config()
        except Exception as e:
            error = e

        v.set_config_name("app")
        v.set_config_type("yaml")
        v.add_config_path(".")

        if v.is_set("APP_CONFIG_PATH"):
            v.add_config_path(v.get("APP_CONFIG_PATH"))

        try:
            v.merge_in_config()
        except Exception as e:
            error = e

        if len(v.all_keys()) == 0:
            logger.error("no configuration found")
            raise FailedLoadingConfigError(error)

        v.set_default("metadata.base_path", "/")
        v.set_default("runner.logger.level", "INFO")
        v.set_default("runner.logger.output_paths", ["stdout"])
        v.set_default("runner.logger.error_output_paths", ["stderr"])

    def initialize_logger(self):
        logger_format = "<green>{time}</green> <level>{extra[context]} {message}</level>"
        output_paths = v.get("runner.logger.output_paths")
        error_output_paths = v.get("runner.logger.error_output_paths")

        logger.remove()  # Remove the pre-configured handler
        for output_path in output_paths:
            if output_path == "stdout":
                output_path = sys.stdout

            logger.add(
                output_path,
                colorize=True,
                format=logger_format,
                backtrace=False,
                diagnose=False,
                level=v.get("runner.logger.level"),
            )
        for error_output_path in error_output_paths:
            if output_path == "stderr":
                output_path = sys.stderr

            logger.add(
                error_output_path,
                colorize=True,
                format=logger_format,
                backtrace=True,
                diagnose=True,
                level="ERROR",
            )

        self.logger = logger.bind(context="[RUNNER CONFIG]")
        self.logger.info("logger initialized")

    async def trigger_runner(self):
        return await TriggerRunner(self.js, self.logger).initialize()

    async def task_runner(self):
        return await TaskRunner(self.js, self.logger).initialize()

    async def exit_runner(self):
        return await ExitRunner(self.js, self.logger).initialize()
